### PR TITLE
REALMC-8135: Implement data-source flag for app create command

### PR DESCRIPTION
**Pull request by nathantfrank**

### CODE REVIEW

File: internal/cloud/realm/app.go [hunk 1]
```diff
@@ -109,7 +109,7 @@ func (c *client) FindApps(filter AppFilter) ([]App, error) {
 
 	var filtered = make([]App, 0, len(apps))
 	for _, app := range apps {
-		if strings.HasPrefix(app.ClientAppID, filter.App) {
+		if strings.HasPrefix(app.ClientAppID, strings.ToLower(filter.App)) {
```
Comment (nathantfrank) [inline internal/cloud/realm/app.go:112]: Noticed the ClientAppID would be all lowercased which would cause name not to be matched

File: internal/cloud/realm/command_test.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+	t.Run("with an active session", func(t *testing.T) {
+		client := newAuthClient(t)
+		groupID := u.CloudGroupID()
+
+		t.Run("should list clusters", func(t *testing.T) {
+			app, err := client.CreateApp(groupID, "commands-test", realm.AppMeta{})
+			assert.Nil(t, err)
+
+			_, err = client.ListClusters(groupID, app.ID)
```
Comment (nathantfrank) [inline internal/cloud/realm/command_test.go:31]: [Q] Is this okay? There could be any number of available clusters or none.

File: internal/commands/app/create_test.go [hunk 3]
```diff
@@ -12,7 +12,6 @@ import (
 	"github.com/10gen/realm-cli/internal/local"
 	"github.com/10gen/realm-cli/internal/utils/test/assert"
 	"github.com/10gen/realm-cli/internal/utils/test/mock"
-
```
Comment (nathantfrank) [inline internal/commands/app/create_test.go:15]: Will be adding this line back in next commit

**Review by nathantfrank — COMMENTED** [review]
First Pass

**Comment by nathantfrank** [comment]
@makesitgo Do we want to have an additional line of output about the datasource if all went well?

File: internal/cloud/realm/command_test.go (hunk 2 above)
Comment (makesitgo) [reply internal/cloud/realm/command_test.go:31]: if im understanding this correctly, we can probably tighten this up via env vars and recommending devs to setup their own test projects to mirror the setup we have for the test user here (similar to how we leverage `u.CloudGroupName()`)

**Comment by makesitgo** [comment]
Yes on the additional line of output...

question: have you tested this out locally?  is the api call at all lengthy?  I ask b/c we (will soon) support spinners, and Im wondering if this process warrants one.

either way, you could have a `01:23:45 UTC INFO  Linked data source` or some type of message similar.

if a spinner's invovlved, the way it works so far is we have a temporary message like:

`:spinner: Linking data source...`

which would then disappear and be replaced with a log message instead:

`01:23:45 UTC INFO  Linked data source`

Happy to elaborate more or demo it locally (you can check it out as well [here](https://github.com/makesitgo/realm-cli/tree/REALMC-7867) -- just no promises that branch is stable and won't see any force pushes)

File: internal/commands/app/create.go [hunk 4]
```diff
@@ -109,56 +112,60 @@ func (cmd *CommandCreate) Handler(profile *cli.Profile, ui terminal.UI) error {
...
+		}
+		serviceName := newApp.Name + "_cluster"
+		serviceConfig := map[string]interface{}{
+			"name": serviceName,
+			"type": "mongodb-atlas",
+			"config": map[string]interface{}{
+				"clusterName":         dsCluster,
+				"readPreference":      "primary",
+				"wireProtocolEnabled": false,
+			},
```
Comment (kpatel71716) [inline internal/commands/app/create.go:138]: [major q] Should we consider data lakes here as well? Data Lakes, support for which is relatively new (mid-late 2020), and Clusters are the two big types of data sources we have. I believe the service configuration is different as well for data lakes so that is something to keep in mind. Though I haven't looked through the master branch yet to see if we added support for data lakes in the cli last year

**Comment by nathantfrank** [comment]
> Yes on the additional line of output...
> 
> question: have you tested this out locally? is the api call at all lengthy? I ask b/c we (will soon) support spinners, and Im wondering if this process warrants one.
> 
> either way, you could have a `01:23:45 UTC INFO Linked data source` or some type of message similar.
> 
> if a spinner's invovlved, the way it works so far is we have a temporary message like:
> 
> `:spinner: Linking data source...`
> 
> which would then disappear and be replaced with a log message instead:
> 
> `01:23:45 UTC INFO Linked data source`
> 
> Happy to elaborate more or demo it locally (you can check it out as well [here](https://github.com/makesitgo/realm-cli/tree/REALMC-7867) -- just no promises that branch is stable and won't see any force pushes)

From my testing against dev there is not a significant time difference (> ~1 sec) when linking a data source vs just app creation.

File: internal/commands/commands.go [hunk 5]
```diff
@@ -74,7 +74,20 @@ Realm app will be exported as well.`,
...
-			{
+			cli.CommandDefinition{
+				Use:         "create",
+				Display:     "app create",
+				Description: "Create a remote Realm app from your current working directory",
+				Help: `Creates a new Realm app in the cloud, and saves your configuration files on 
+a new disk. This command will create a new directory for your project. 
+To initialize a Realm app in your current working directory, use ‘app init’.
+
+You can specify a ‘From’ input to create a Realm app from either a template,
```
Comment (makesitgo) [inline internal/commands/commands.go:85]: i know this is from the doc, but we should tweak this a little to avoid mentioning `templates` (as thats not supported/implemented yet).  I also think using `'--from'` instead of `'From'` makes more sense in this context, thoughts?

File: internal/cloud/realm/command_test.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+		_, err := client.ListClusters(primitive.NewObjectID().Hex(), primitive.NewObjectID().Hex())
+		assert.Equal(t, realm.ErrInvalidSession{}, err)
+	})
+
+	t.Run("with an active session", func(t *testing.T) {
+		client := newAuthClient(t)
+		groupID := u.CloudGroupID()
+
+		t.Run("should list clusters", func(t *testing.T) {
+			app, err := client.CreateApp(groupID, "commands-test", realm.AppMeta{})
```
Comment (makesitgo) [inline internal/cloud/realm/command_test.go:28]: we should use the `setupApp` helper here and `defer teardown()` with it

File: internal/commands/app/create.go [hunk 7]
```diff
@@ -101,70 +104,66 @@ func (cmd *CommandCreate) Handler(profile *cli.Profile, ui terminal.UI, clients
...
-		},
-		{
-			"Info":    "Realm UI",
-			"Details": fmt.Sprintf("%s/groups/%s/apps/%s/dashboard", profile.RealmBaseURL(), newApp.GroupID, newApp.ID),
-		},
-		{
-			"Info":    "Check out your app",
-			"Details": fmt.Sprintf("cd ./%s && realm-cli app describe", newApp.Name),
-		},
+	rows := make([]map[string]interface{}, 0, rowCount)
```
Comment (makesitgo) [inline internal/commands/app/create.go:158]: you can just let this be `make([]map[string]interface{}, 0, 4)`

File: internal/commands/app/create.go [hunk 8]
```diff
@@ -101,70 +104,66 @@ func (cmd *CommandCreate) Handler(profile *cli.Profile, ui terminal.UI, clients
...
+			return err
+		}
+		rowCount++
+	}
 
 	if err := clients.Realm.Import(
 		newApp.GroupID,
 		newApp.ID,
-		loadedApp,
+		loadedApp.AppData,
```
Comment (makesitgo) [inline internal/commands/app/create.go:152]: was this change needed?

File: internal/commands/app/create_inputs.go [hunk 9]
```diff
@@ -18,10 +18,9 @@ var (
 	flagDirectoryShort = "c"
 	flagDirectoryUsage = "the directory to create your new Realm app, defaults to Realm app name"
 
-	// TODO(REALMC-8135): Implement data-source flag for app create command
-	// flagDataSource      = "data-source"
-	// flagDataSourceShort = "s"
-	// flagDataSourceUsage = "atlas cluster to back your Realm app, defaults to first available"
+	flagDataSource      = "data-source"
+	flagDataSourceShort = "s"
+	flagDataSourceUsage = "atlas cluster to back your Realm app, defaults to first available"
```
Comment (makesitgo) [inline internal/commands/app/create_inputs.go:23]: `include to link an Atlas cluster to your Realm app, defaults to first available`

File: internal/commands/app/create_inputs.go [hunk 10]
```diff
@@ -88,29 +85,28 @@ func (i *createInputs) resolveDirectory(wd string) (string, error) {
 		return "", appErr
 	}
 	if appOK {
-		return "", fmt.Errorf("%s is inside or is a Realm app directory", dir)
+		return "", fmt.Errorf("%s is inside or is a Realm app directory", fullPath)
```
Comment (makesitgo) [inline internal/commands/app/create_inputs.go:88]: ideally we use some form of `errProjectExists` here, so that its consistent b/w `app init` and `app create`

File: internal/commands/app/create.go [hunk 11]
```diff
@@ -101,70 +104,66 @@ func (cmd *CommandCreate) Handler(profile *cli.Profile, ui terminal.UI, clients
...
-	// if idx := strings.LastIndex(zipPkgName, "_"); idx != -1 {
-	// 	zipPkgName = zipPkgName[:idx]
-	// }
-	// zipPkgPath := filepath.Join(dir, zipPkgName) // Is this the correct path
-	// if writeErr := local.WriteZip(zipPkgPath, zipPkg); writeErr != nil {
-	// 	return writeErr
-	// }
+	rowCount := 3
+	if cmd.inputs.DataSource != "" {
+		dsCluster, err := cmd.inputs.resolveDataSource(clients.Realm, groupID, newApp.ID)
```
Comment (makesitgo) [inline internal/commands/app/create.go:109]: if we keep this function name, id rename this `var` to `dataSource`.  otherwise I'd consider renaming this to `resolveClusterName` to be more specific about what its doing (for if and when data lakes become a thing)

File: internal/commands/app/create_inputs.go [hunk 12]
```diff
@@ -88,29 +85,28 @@ func (i *createInputs) resolveDirectory(wd string) (string, error) {
...
-// 		return "", fmt.Errorf("Unable to find any available cluster for Group ID %s", groupID)
-// 	}
-// 	return dsCluster, nil
-// }
+func (i *createInputs) resolveDataSource(client realm.Client, groupID, appID string) (string, error) {
+	clusters, err := client.ListClusters(groupID, appID)
+	if err != nil {
+		return "", err
+	}
+	var dsCluster string
```
Comment (makesitgo) [inline internal/commands/app/create_inputs.go:98]: [nit] rename this to `clusterName`

File: internal/commands/app/create_inputs.go [hunk 13]
```diff
@@ -88,29 +85,28 @@ func (i *createInputs) resolveDirectory(wd string) (string, error) {
...
+		if (i.DataSource == "" && cluster.State == "IDLE") || i.DataSource == cluster.Name {
+			dsCluster = cluster.Name
+			break
+		}
+	}
+	if dsCluster == "" {
+		if i.DataSource != "" {
+			return "", fmt.Errorf("unable to find the %s cluster", i.DataSource)
+		}
+		return "", fmt.Errorf("unable to find any available cluster for Project %s", groupID)
```
Comment (makesitgo) [inline internal/commands/app/create_inputs.go:109]: thoughts on simplifying this and just returning `errors.New("failed to find Atlas cluster")` here?  in the context of running this command, the user probably already know the name they specified

File: internal/commands/app/create.go [hunk 14]
```diff
@@ -101,70 +104,66 @@ func (cmd *CommandCreate) Handler(profile *cli.Profile, ui terminal.UI, clients
...
+			"config": map[string]interface{}{
+				"clusterName":         dsCluster,
+				"readPreference":      "primary",
+				"wireProtocolEnabled": false,
+			},
+		}
+		var dsPath string
+		switch loadedApp.ConfigVersion() {
+		case realm.AppConfigVersion20210101:
+			dsPath = fmt.Sprintf("data_sources/%s/config.json", serviceName)
```
Comment (makesitgo) [inline internal/commands/app/create.go:126]: use `filepath.Join` for these types of things:

`filepath.Join(local.NameDataSources, serviceName, local.FileConfig.String()`

(similarly for the below case as well)

File: internal/commands/app/create.go [hunk 15]
```diff
@@ -101,70 +104,66 @@ func (cmd *CommandCreate) Handler(profile *cli.Profile, ui terminal.UI, clients
...
+		serviceConfig := map[string]interface{}{
+			"name": serviceName,
+			"type": "mongodb-atlas",
+			"config": map[string]interface{}{
+				"clusterName":         dsCluster,
+				"readPreference":      "primary",
+				"wireProtocolEnabled": false,
+			},
+		}
+		var dsPath string
```
Comment (makesitgo) [inline internal/commands/app/create.go:123]: [meganit] `path` or `serviceConfigPath` here

File: internal/commands/app/create.go [hunk 16]
```diff
@@ -101,70 +104,66 @@ func (cmd *CommandCreate) Handler(profile *cli.Profile, ui terminal.UI, clients
...
+		data, err := json.MarshalIndent(serviceConfig, local.ExportedJSONPrefix, local.ExportedJSONIndent)
+		if err != nil {
+			return err
+		}
+		err = local.WriteFile(filepath.Join(dir, dsPath), 0666, bytes.NewReader(data))
+		if err != nil {
+			return err
+		}
+		err = loadedApp.Load()
+		if err != nil {
```
Comment (makesitgo) [inline internal/commands/app/create.go:143]: at least in-line this `err` (i'd do it above too, even if it means splitting the args onto new lines, but your call there)

File: internal/commands/app/create.go [hunk 17]
```diff
@@ -101,70 +104,66 @@ func (cmd *CommandCreate) Handler(profile *cli.Profile, ui terminal.UI, clients
...
+	rows := make([]map[string]interface{}, 0, rowCount)
+	rows = append(rows, map[string]interface{}{"Info": "Client App ID", "Details": newApp.ClientAppID})
+	rows = append(rows, map[string]interface{}{"Info": "Realm Directory", "Details": dir})
+	rows = append(rows, map[string]interface{}{"Info": "Realm UI", "Details": fmt.Sprintf("%s/groups/%s/apps/%s/dashboard", profile.RealmBaseURL(), newApp.GroupID, newApp.ID)})
+	if cmd.inputs.DataSource != "" {
+		rows = append(rows, map[string]interface{}{"Info": "Data Source", "Details": cmd.inputs.DataSource})
 	}
 
 	ui.Print(terminal.NewTableLog("Successfully created app", headers, rows...))
+	ui.Print(terminal.NewDebugLog("Check out your app: cd ./%s && realm-cli app describe", cmd.inputs.Directory))
```
Comment (makesitgo) [inline internal/commands/app/create.go:167]: lets create `terminal.NewFollowupLog` which takes in `message string` and `items ...interface{}`, and will otherwise replace what we have for `NewReferenceLinksLog`/`NewSuggestedCommandsLog`.  You could then also use that here

File: internal/commands/app/create_inputs.go [hunk 18]
```diff
@@ -88,29 +85,28 @@ func (i *createInputs) resolveDirectory(wd string) (string, error) {
...
-// 	}
-// 	if dsCluster == "" {
-// 		if i.DataSource != "" {
-// 			return "", fmt.Errorf("Unable to find the %s cluster", i.DataSource)
-// 		}
-// 		return "", fmt.Errorf("Unable to find any available cluster for Group ID %s", groupID)
-// 	}
-// 	return dsCluster, nil
-// }
+func (i *createInputs) resolveDataSource(client realm.Client, groupID, appID string) (string, error) {
```
Comment (makesitgo) [inline internal/commands/app/create_inputs.go:93]: do we have tests for this here?  probably somewhat implicitly through the `Handler` tests, but can we write a few unit tests for this to make sure we're returning the errors we expect, and also matching on the `"IDLE"`/`cluster.Name` logic?

**Review by makesitgo — COMMENTED** [review]
lookin real good so far... nothing major just a bunch of questions/nits

File: internal/commands/commands.go (hunk 5 above)
Comment (nathantfrank) [reply internal/commands/commands.go:85]: agree

File: internal/commands/app/create_inputs.go (hunk 13 above)
Comment (nathantfrank) [reply internal/commands/app/create_inputs.go:109]: I agree, though I do like the difference in error between failing to find one they specified vs any available cluster

File: internal/commands/app/create.go (hunk 8 above)
Comment (nathantfrank) [reply internal/commands/app/create.go:152]: Not required, but it does read a little more cleanly in the context of the `Import` parameter `appData`

Comment (makesitgo) [reply internal/commands/app/create.go:139]: talked offline, but to close the loop here, this was need (and the source of our Diff/Import bug I was seeing earlier today)

File: internal/commands/app/create_inputs.go (hunk 13 above)
Comment (makesitgo) [reply internal/commands/app/create_inputs.go]: to close the loop here as well, this scenario does not exist with the current implementation.

should we ever support these bool/string hybrid-type flags, we can revisit this

File: internal/commands/app/create.go (hunk 4 above)
Comment (makesitgo) [reply internal/commands/app/create.go]: this flag is new functionality, so there was no prior support for this.  this is a great point though... lets tackle that in a separate jira though

File: internal/commands/app/create_inputs.go [hunk 19]
```diff
@@ -83,34 +81,39 @@ func (i *createInputs) resolveDirectory(wd string) (string, error) {
...
-// 	}
-// 	if dsCluster == "" {
-// 		if i.DataSource != "" {
-// 			return "", fmt.Errorf("Unable to find the %s cluster", i.DataSource)
-// 		}
-// 		return "", fmt.Errorf("Unable to find any available cluster for Group ID %s", groupID)
-// 	}
-// 	return dsCluster, nil
-// }
+func (i *createInputs) resolveDataSource(client realm.Client, groupID, appID string) (map[string]interface{}, error) {
```
Comment (makesitgo) [inline internal/commands/app/create_inputs.go:94]: instead of returning `map[string]interface{}` here, let's just a struct for this (so we can avoid the `["name"]` accessor and use `.Name` instead)

just keep it privately declared here for now (`type dataSource struct { ... }`).  its tempting to say put this in `local` and we can use it as part of the `ServiceStructure`, but I still want to try and keep that `local` structure stuff as generic as possible

File: internal/commands/app/create.go [hunk 20]
```diff
@@ -101,70 +103,55 @@ func (cmd *CommandCreate) Handler(profile *cli.Profile, ui terminal.UI, clients
...
-	// 	return writeErr
-	// }
-
-	if err := clients.Realm.Import(
+	if cmd.inputs.DataSource != "" {
+		dataSource, err := cmd.inputs.resolveDataSource(clients.Realm, groupID, newApp.ID)
+		if err != nil {
+			return err
+		}
+		dataSourceName := fmt.Sprintf("%v", dataSource["name"])
```
Comment (makesitgo) [inline internal/commands/app/create.go:111]: can get rid of this once we convert to a `struct` here

File: internal/commands/app/create.go [hunk 21]
```diff
@@ -101,70 +103,55 @@ func (cmd *CommandCreate) Handler(profile *cli.Profile, ui terminal.UI, clients
...
+		var path string
+		switch loadedApp.ConfigVersion() {
+		case realm.AppConfigVersion20210101:
```
Comment (makesitgo) [inline internal/commands/app/create.go:114]: let's make this the default case/behavior and the potential error here

File: internal/commands/app/errors.go [hunk 22]
```diff
@@ -1,7 +1,14 @@
 package app
 
-type errProjectExists struct{}
+type errProjectExists struct {
+	details string
```
Comment (makesitgo) [inline internal/commands/app/errors.go:4]: any objections to making this `path string`, and then using
```go
var suffix string
if err.path != "" {
  suffix = " at "+err.path
}
return "a project already exists" + suffix
```
for the error string?

File: internal/commands/app/create_inputs.go [hunk 23]
```diff
@@ -83,34 +81,39 @@ func (i *createInputs) resolveDirectory(wd string) (string, error) {
...
-	_, appOK, appErr := local.FindApp(fullPath)
-	if appErr != nil {
-		return "", appErr
+	_, appOK, err := local.FindApp(fullPath)
+	if err != nil {
+		return "", err
 	}
 	if appOK {
-		return "", fmt.Errorf("%s is inside or is a Realm app directory", dir)
+		return "", errProjectExists{details: fmt.Sprintf("%s is inside or is a Realm app directory", fullPath)}
```
Comment (makesitgo) [inline internal/commands/app/create_inputs.go:89]: if we make the switch to `path string` for this error struct, this can be just `errProjectExists{"./"+i.Directory}`

File: internal/commands/app/create_inputs_test.go [hunk 24]
```diff
@@ -193,6 +193,69 @@ func TestAppCreateInputsResolveDirectory(t *testing.T) {
 		dir, err := inputs.resolveDirectory(profile.WorkingDirectory)
 
 		assert.Equal(t, "", dir)
-		assert.Equal(t, fmt.Errorf("%s is inside or is a Realm app directory", specifiedDir), err)
+		assert.Equal(t, errProjectExists{details: fmt.Sprintf("%s is inside or is a Realm app directory", fullDir)}, err)
+	})
+}
+
+func TestAppCreateInputsResolveDataSource(t *testing.T) {
+	t.Run("should return data source config of an provided cluster", func(t *testing.T) {
```
Comment (makesitgo) [inline internal/commands/app/create_inputs_test.go:201]:
```suggestion
	t.Run("should return data source config of a provided cluster", func(t *testing.T) {
```

**Review by makesitgo — APPROVED** [review]
minor things left, but this is good to go

🆒  beans
